### DISCUSSION TITLE
StoryBook Interaction testing - Jest Dom helpers

### DISCUSSION THREAD

**Discussion by jnahumphreys** [storybookjs/storybook#16858] [Help] [answered]
Hi, great work on the 6.4 interaction feature. I'm migrating a component test from a JSDOM environment (@testing-library/react & Jest) to the new StoryBook play API.

Does `@storybook/testing-library` expose `@testing-library/jest-dom` type helpers. e.g `toBeDisabled()`, `toBeInTheDocument()` etc. If not does the team have any workaround / suggestions they can recommend?

Thanks in advance

**Comment by yannbf** [comment]
**edit:** This is not necessary anymore, see [this answer](https://github.com/storybookjs/storybook/discussions/16858#discussioncomment-2196253) instead

Hey @BinaryJim great question! We are currently experimenting with a way to automatically provide the helpers, but we are not there yet. 

You might be able to get the functionality by adding this to your `preview.js` file
```js
// .storybook/preview
import { expect } from '@storybook/jest'
import * as matchers from '@testing-library/jest-dom/matchers'
expect.extend(matchers)
```

However you won't get the types. For that you'd need to do module augmentation, I believe. I'd have to get back to you on that!

**Reply by jnahumphreys** [reply]
Thanks for the prompt reply @yannbf - I'll give that a try and keep an eye out for an official workaround at some point.

**Comment by yannbf** [comment, marked as answer]
Hey everyone and @BinaryJim I'm happy to announce that this feature has been implemented and you can access it by upgrading your `@storybook/jest` dependency to `0.0.9`! You might have to delete your package-lock or yarn.lock files as this library depends on `@storybook/expect` and should use its latest version `27.5.2-0`.

With this there is no need to do any workaround, it should just work out of the box!

**Reply by jnahumphreys** [reply]
Awesome news! Thanks for the update and the team’s efforts on this feature 👍

**Reply by jnahumphreys** [reply]
@yannbf - Finally getting round to using the JestDom utils.

**Example within my Story play function**

```javascript
expect(canvas.getByRole('button')).toBeDisabled()
```
How do we call theses helpers

**Reply by yannbf** [reply]
Hey Jim! That should just work. Is that not working for you? Make sure you are using `@storybook/jest` and `@storybook/testing-library`.

**Reply by jnahumphreys** [reply]
Hey.

Unfortunately not, here's an extract from my Textarea story:

```typescript
import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { expect } from "@storybook/jest";
import { within, userEvent } from "@storybook/testing-library";
import { Textarea } from "./textarea";

export default {
  title: "Components/Textarea",
  component: Textarea,
  argTypes: {
    onChange: {
      action: "Input value changed",
    },
  },
} as ComponentMeta<typeof Textarea>;

const Template: ComponentStory<typeof Textarea> = (args) => (
  <Textarea {...args} />
);

export const Disabled = Template.bind({});

Disabled.args = {
  disabled: true,
  defaultValue: "Hello World",
};

Disabled.play = async ({ canvasElement }) => {
  const canvas = within(canvasElement);
  const textAreaElement = canvas.getByRole("textbox") as HTMLTextAreaElement;
  expect(window.getComputedStyle(textAreaElement).cursor).toBe("not-allowed");
  expect(textAreaElement).toBeDisabled();
};
```

VSCode is flagging this type error when calling `toBeDisabled()`

```typescript
expect(textAreaElement).toBeDisabled();
// Property 'toBeDisabled' does not exist on type 'JestMatchers<HTMLTextAreaElement>'.ts(2339)
```

And storybook fails to compile with this output to my console

```
ERROR in components/textarea/textarea.stories.tsx:87:27
TS2339: Property 'toBeDisabled' does not exist on type 'JestMatchers<HTMLTextAreaElement>'.
    85 |   expect(textAreaElement.disabled).toBe(true);
    86 |   expect(window.getComputedStyle(textAreaElement).cursor).toBe("not-allowed");
  > 87 |   expect(textAreaElement).toBeDisabled();
       |                           ^^^^^^^^^^^^
    88 | };
    89 |
    90 | export const ReadOnly = Template.bind({});
```

Not quite sure what's up here 🤔 Any ideas?

**Reply by FinalDoom** [reply]
I'm am able to use this with typescript, @storybook/testing-library 0.0.13, @storybook/jest 0.0.10, @testing-library/jest-dom 5.16.5, and @storybook/expect 27.5.2-0, but VSCode gives errors that the method doesn't exist. Maybe just the typings aren't being extended correctly?

**Reply by GraemeF** [reply]
That seems to be the case - the functions are there at runtime but TypeScript compiler can't see them.

**Reply by pswai** [reply]
@FinalDoom @GraemeF I temporarily worked around this by adding `@types/testing-library__jest-dom` to my tsconfig.json, not sure if there is a better way

**Reply by yannbf** [reply]
Hey everyone! This is likely an issue with package managers.
The type augmentation comes from `@storybook/jest`, which depends on `@testing-library/jest-dom`, which depends on `@types/testing-library__jest-dom`. That package should be hoisted, and, as a result, you should have a `node_modules/@types/testing-library__jest-dom` directory. If you don't, then the types won't exist.

For instance, if you have a `yarn` project, that happens automatically.
If you're using `pnpm`, [@types are not hoisted automatically](https://github.com/pnpm/pnpm/issues/4920), you will have to add a `.npmrc` file with:
```
public-hoist-pattern[]=@types*
```

Please let me know which package managers you are using (and if it is a monorepo) and if you can solve the problem with hoisting settings!

Otherwise, installing `@types/testing-library__jest-dom` as a direct dependency should have the same effect, though it's not great that you have to add the extra dependency.

**Reply by FinalDoom** [reply]
That is probably it for me. Thank you for the fix: I'm using `pnpm` so that should fix it for me. Lots of nuance to learn/adjust to for a faster manager.

Edit: Yup, that fixed it after `pnpm i`. Nice!

**Reply by bryanjtc** [reply]
Is there a workaround for yarn pnp without installing @types/testing-library__jest-dom as a direct dependency?

**Reply by yannbf** [reply]
@bryanjtc I have no idea (but would love to hear if you find out). It's unfortunately an issue with Testing Library itself, even if you were not using it with Storybook. You can see here:
https://github.com/testing-library/react-testing-library/issues/620
https://github.com/testing-library/jest-dom/issues/314
